// eslint-plugin-reactintl/lib/index.js
'use strict';

var _defaultMessage = require('./rules/default-message/default-message');

var _defaultMessage2 = _interopRequireDefault(_defaultMessage);

var _containsHardcodedCopy = require('./rules/contains-hardcoded-copy/contains-hardcoded-copy');

var _containsHardcodedCopy2 = _interopRequireDefault(_containsHardcodedCopy);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @fileoverview Ensures react-intl components have a defaultMessage supplied
 * @author Dan Pavitt
 */

module.exports = {
    rules: {
        'default-message': _defaultMessage2.default,
        'contains-hardcoded-copy': _containsHardcodedCopy2.default
    },
    configs: {
        recommended: {
            parserOptions: {
                ecmaFeatures: {
                    jsx: true
                }
            },
            rules: {
                'default-message': 'warning',
                'contains-hardcoded-copy': 'off'
            }
        },
        strict: {
            parserOptions: {
                ecmaFeatures: {
                    jsx: true
                }
            },
            rules: {
                'default-message': 'warning',
                'contains-hardcoded-copy': 'off'
            }
        }
    }
};
